In [59]:
import pandas as pd  #Importing necessary packages

In [61]:
#importing .csv file to work on
survey_responses = pd.read_csv('/Users/anthonydaniels/Desktop/Udacity Data Science Study/Udacity Project 1/Udacity Project 1 csv dataset 1.csv')

In [63]:
survey_responses.head()  #look at first 5 lines of code



ResponseId                      MainBranch                 Age  \
0           1  I am a developer by profession  Under 18 years old   
1           2  I am a developer by profession     35-44 years old   
2           3  I am a developer by profession     45-54 years old   
3           4           I am learning to code     18-24 years old   
4           5  I am a developer by profession     18-24 years old   

            Employment RemoteWork   Check  \
0  Employed, full-time     Remote  Apples   
1  Employed, full-time     Remote  Apples   
2  Employed, full-time     Remote  Apples   
3   Student, full-time        NaN  Apples   
4   Student, full-time        NaN  Apples   

                                    CodingActivities  \
0                                              Hobby   
1  Hobby;Contribute to open-source projects;Other...   
2  Hobby;Contribute to open-source projects;Other...   
3                                                NaN   
4                                                NaN   

                                             EdLevel  \
0                          Primary/elementary school   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3  Some college/university study without earning ...   
4  Secondary school (e.g. American high school, G...   

                                           LearnCode  \
0                             Books / Physical media   
1  Books / Physical media;Colleague;On the job tr...   
2  Books / Physical media;Colleague;On the job tr...   
3  Other online resources (e.g., videos, blogs, f...   
4  Other online resources (e.g., videos, blogs, f...   

                                     LearnCodeOnline  ... JobSatPoints_6  \
0                                                NaN  ...            NaN   
1  Technical documentation;Blogs;Books;Written Tu...  ...            0.0   
2  Technical documentation;Blogs;Books;Written Tu...  ...            NaN   
3  Stack Overflow;How-to videos;Interactive tutorial  ...            NaN   
4  Technical documentation;Blogs;Written Tutorial...  ...            NaN   

  JobSatPoints_7 JobSatPoints_8 JobSatPoints_9 JobSatPoints_10  \
0            NaN            NaN            NaN             NaN   
1            0.0            0.0            0.0             0.0   
2            NaN            NaN            NaN             NaN   
3            NaN            NaN            NaN             NaN   
4            NaN            NaN            NaN             NaN   

  JobSatPoints_11           SurveyLength SurveyEase ConvertedCompYearly JobSat  
0             NaN                    NaN        NaN                 NaN    NaN  
1             0.0                    NaN        NaN                 NaN    NaN  
2             NaN  Appropriate in length       Easy                 NaN    NaN  
3             NaN               Too long       Easy                 NaN    NaN  
4             NaN              Too short       Easy                 NaN    NaN  

[5 rows x 114 columns]

In [65]:
survey_responses.describe() #descriptive statistics of file

ResponseId      CompTotal       WorkExp  JobSatPoints_1  \
count  65437.000000   3.374000e+04  29658.000000    29324.000000   
mean   32719.000000  2.963841e+145     11.466957       18.581094   
std    18890.179119  5.444117e+147      9.168709       25.966221   
min        1.000000   0.000000e+00      0.000000        0.000000   
25%    16360.000000   6.000000e+04      4.000000        0.000000   
50%    32719.000000   1.100000e+05      9.000000       10.000000   
75%    49078.000000   2.500000e+05     16.000000       22.000000   
max    65437.000000  1.000000e+150     50.000000      100.000000   

       JobSatPoints_4  JobSatPoints_5  JobSatPoints_6  JobSatPoints_7  \
count    29393.000000    29411.000000    29450.000000     29448.00000   
mean         7.522140       10.060857       24.343232        22.96522   
std         18.422661       21.833836       27.089360        27.01774   
min          0.000000        0.000000        0.000000         0.00000   
25%          0.000000        0.000000        0.000000         0.00000   
50%          0.000000        0.000000       20.000000        15.00000   
75%          5.000000       10.000000       30.000000        30.00000   
max        100.000000      100.000000      100.000000       100.00000   

       JobSatPoints_8  JobSatPoints_9  JobSatPoints_10  JobSatPoints_11  \
count    29456.000000    29456.000000     29450.000000     29445.000000   
mean        20.278165       16.169432        10.955713         9.953948   
std         26.108110       24.845032        22.906263        21.775652   
min          0.000000        0.000000         0.000000         0.000000   
25%          0.000000        0.000000         0.000000         0.000000   
50%         10.000000        5.000000         0.000000         0.000000   
75%         25.000000       20.000000        10.000000        10.000000   
max        100.000000      100.000000       100.000000       100.000000   

       ConvertedCompYearly        JobSat  
count         2.343500e+04  29126.000000  
mean          8.615529e+04      6.935041  
std           1.867570e+05      2.088259  
min           1.000000e+00      0.000000  
25%           3.271200e+04      6.000000  
50%           6.500000e+04      7.000000  
75%           1.079715e+05      8.000000  
max           1.625660e+07     10.000000

In [67]:
survey_responses.isna().sum() #check for nulls and count them

ResponseId                 0
MainBranch                 0
Age                        0
Employment                 0
RemoteWork             10631
                       ...  
JobSatPoints_11        35992
SurveyLength            9255
SurveyEase              9199
ConvertedCompYearly    42002
JobSat                 36311
Length: 114, dtype: int64

In [69]:
survey_responses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Columns: 114 entries, ResponseId to JobSat
dtypes: float64(13), int64(1), object(100)
memory usage: 56.9+ MB


We have 114 columns/features in our dataset.  We need to start the cleaning process.
Our project is going to attempt create a model that predicts those who will create a career which will be defined as those with over 10 years work experience.

In [71]:
# eliminate columns that we don't want to be taken into account for our model
# we have a column 'check' that appears to have responses all 'apples'.  check to see if there are any other responses

# Check for non-'apples' responses in the 'check' column
non_apples = survey_responses[survey_responses['Check'] != 'Apples']

# Display results
if len(non_apples) > 0:
    print(f"Found {len(non_apples)} rows with non-'apples' responses:")
    print(non_apples[['Check']])  # Show only the 'check' column
else:
    print("All responses in 'Check' column are 'Apples'.")

All responses in 'Check' column are 'Apples'.


Based on the fact that all of the responses are 'Apples' under the 'Check' column, we can assume that it is a nonimpactful column since it applies to everybody in the survey

In [73]:
#code to remove undesired columns from file
survey_responses.drop(columns=['Check', 'OrgSize', 'PurchaseInfluence', 'BuyNewTool', 'BuildvsBuy', 'TechEndorse', 'Currency','CompTotal', 'LanguageAdmired', 'DatabaseAdmired', 'PlatformAdmired', 'WebframeAdmired', 'EmbeddedAdmired', 'MiscTechWantToWorkWith', 'MiscTechAdmired', 'ToolsTechAdmired', 'NEWCollabToolsWantToWorkWith', 'NEWCollabToolsAdmired', 'OfficeStackAsyncWantToWorkWith', 'OfficeStackAsyncAdmired', 'OfficeStackSyncWantToWorkWith', 'OfficeStackSyncAdmired', 'AISearchDevWantToWorkWith', 'AISearchDevAdmired', 'SurveyLength', 'SurveyEase', 'ConvertedCompYearly'], inplace=True)

In [75]:
survey_responses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 87 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      65437 non-null  int64  
 1   MainBranch                      65437 non-null  object 
 2   Age                             65437 non-null  object 
 3   Employment                      65437 non-null  object 
 4   RemoteWork                      54806 non-null  object 
 5   CodingActivities                54466 non-null  object 
 6   EdLevel                         60784 non-null  object 
 7   LearnCode                       60488 non-null  object 
 8   LearnCodeOnline                 49237 non-null  object 
 9   TechDoc                         40897 non-null  object 
 10  YearsCode                       59869 non-null  object 
 11  YearsCodePro                    51610 non-null  object 
 12  DevType                         

In [77]:
WorkExp_NA = survey_responses['WorkExp'].isna().sum()  #determine if there are NA to be dealth with in WorkExp column
print (WorkExp_NA)

35779


In [79]:
survey_responses['Career'] = survey_responses['WorkExp'].fillna(0) >=10   #creates new column named Career and assigns False to everything w/NA or Less than 10 years of work experience

In [81]:
print(survey_responses[['WorkExp','Career']].head()) ##checks WorkExp vs new Career column

   WorkExp  Career
0      NaN   False
1     17.0    True
2      NaN   False
3      NaN   False
4      NaN   False


In [83]:
survey_responses['WorkExp'] = survey_responses['WorkExp'].fillna(0)  #replaces all 'NaN' rows under 'WorkExp' to an assumed work experience of 0

In [85]:
survey_responses['MainBranch'] = survey_responses['MainBranch'].fillna('No_Experience') ### Replaces all 'Nan' responses in 'MainBranch' column to 'No_Experience'


In [87]:
###Converting Responses under 'MainBranch' column to individual columns with Boolean responses

Main_Branch_choices = [
    'I am a developer by profession',
    'I am not primarily a developer, but I write code sometimes as part of my work/studies',
    'I used to be a developer by profession, but no longer am',
    'I am learning to code',
    'I code primarily as a hobby',
    'No_Experience'
]

# Create boolean columns correctly
for option in Main_Branch_choices:
    # Improved column name cleaning
    col_name = (
        option.lower()  # Convert to lowercase
        .replace(' ', '_')
        .replace('/', '_')
        .replace(',', '')
        .replace("'", "")  # Remove apostrophes
    )
    # Compare against the current option only (not the whole list)
    survey_responses[f"Branch_{col_name}"] = (survey_responses['MainBranch'] == option)

In [89]:
survey_responses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 94 columns):
 #   Column                                                                                       Non-Null Count  Dtype  
---  ------                                                                                       --------------  -----  
 0   ResponseId                                                                                   65437 non-null  int64  
 1   MainBranch                                                                                   65437 non-null  object 
 2   Age                                                                                          65437 non-null  object 
 3   Employment                                                                                   65437 non-null  object 
 4   RemoteWork                                                                                   54806 non-null  object 
 5   CodingActivities                

In [91]:
survey_responses = survey_responses.drop(columns=['MainBranch']) #Removes column 'MainBranch'

In [93]:
## Deal w/age object column which has ranges vs actual ages.  convert to integer ages in new 'age_numeric' column
## To avoid too many columns, will use median age points to create integer based row

# Create age mapping dictionary (using midpoint of each range)
age_mapping = {
    'Under 18 years old': 17,
    '18-24 years old': 21,  # (18+24)/2
    '25-34 years old': 29.5,
    '35-44 years old': 39.5,
    '45-54 years old': 49.5,
    '55-64 years old': 59.5,
    '65 years or older': 70,
    'Prefer not to say': None  
}

survey_responses['age_numeric'] = survey_responses['Age'].map(age_mapping)

In [44]:
# Deal w/No Responses or NaN data points in newly created 'age_numeric' column
median_age = survey_responses['age_numeric'].median()
survey_responses['age_numeric'] = survey_responses['age_numeric'].fillna(median_age)

In [95]:
survey_responses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 94 columns):
 #   Column                                                                                       Non-Null Count  Dtype  
---  ------                                                                                       --------------  -----  
 0   ResponseId                                                                                   65437 non-null  int64  
 1   Age                                                                                          65437 non-null  object 
 2   Employment                                                                                   65437 non-null  object 
 3   RemoteWork                                                                                   54806 non-null  object 
 4   CodingActivities                                                                             54466 non-null  object 
 5   EdLevel                         

In [97]:
survey_responses = survey_responses.drop(columns=['Age'])

In [99]:
### Deal w/responses from 'Employment' column
### Will create boolean columns consistent with other responses
## Strip leading and trailing spaces from responses

survey_responses['Employment'] = survey_responses['Employment'].str.strip()

In [101]:
## create employment mapping dictionary to reference in loopin column creation

employment_mapping = {
    'Employed, full-time': 'employed_full_time',
    'Employed, part-time': 'employed_part_time',
    'Independent contractor, freelancer, or self-employed': 'self_employed',
    'Not employed, but looking for work': 'unemployed_seeking',
    'Not employed, and not looking for work': 'unemployed_not_seeking',
    'Student, full-time': 'student_full_time',
    'Student, part-time': 'student_part_time',
    'Retired': 'retired',
    'I prefer not to say': 'prefer_not_to_say'
}

# Create columns boolean in nature
for status, col_name in employment_mapping.items():
    survey_responses[f'{col_name}'] = (survey_responses['Employment'] == status)

In [103]:
survey_responses.info(100)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 102 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    Employment                                                                                   object 
 2    RemoteWork                                                                                   object 
 3    CodingActivities                                                                             object 
 4    EdLevel                                                                                      object 
 5    LearnCode                                                                                    object 
 6    LearnCodeOnline             

In [107]:
survey_responses.info(100)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 102 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    Employment                                                                                   object 
 2    RemoteWork                                                                                   object 
 3    CodingActivities                                                                             object 
 4    EdLevel                                                                                      object 
 5    LearnCode                                                                                    object 
 6    LearnCodeOnline             

In [109]:
### Remove column 'Employment'
survey_responses = survey_responses.drop(columns=['Employment'])

In [111]:
survey_responses.info(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 101 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    RemoteWork                                                                                   object 
 2    CodingActivities                                                                             object 
 3    EdLevel                                                                                      object 
 4    LearnCode                                                                                    object 
 5    LearnCodeOnline                                                                              object 
 6    TechDoc                     

In [113]:
## Deal with 'RemoteWork' column
## Address NA's in column
## Convert 'NAs' to "Remote Not Specified

survey_responses['RemoteWork'] = survey_responses['RemoteWork'].fillna('Remote Not specified')

In [115]:
### Create new boolean columns for each of the response options under'RemoteWork'

remote_responses = [
    'Remote',
    'In-person',
    'Hybrid (some remote, some in-person)',
    'Remote Not Specified']

for responses in remote_responses:
    # Clean mode name for column naming
    col_name = responses.lower()\
        .replace(' ', '_')\
        .replace('(', '')\
        .replace(')', '')\
        .replace('-', '')
    
    # Create boolean column
    survey_responses[f'{col_name}'] = (survey_responses['RemoteWork'] == responses)



In [117]:
survey_responses.info(150)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 105 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    RemoteWork                                                                                   object 
 2    CodingActivities                                                                             object 
 3    EdLevel                                                                                      object 
 4    LearnCode                                                                                    object 
 5    LearnCodeOnline                                                                              object 
 6    TechDoc                     

In [121]:
survey_responses = survey_responses.drop(columns=['RemoteWork'])

Deal With 'CodingActivities' column

In [123]:
## create employment mapping dictionary to reference in loopin column creation

afterwork_mapping = {
    'Hobby': 'hobby',
    'Freelance/contract work': 'freelance_contract',
    'Independent contractor, freelancer, or self-employed': 'self_employed',
    'Contribute to open-source projects': 'open_source_contributor',
    'Bootstrapping a business': 'business_bootstrapper',
    'School or academic work': 'school',
    'Professional development or self-paced learning from online courses': 'online_professional_development',
    'I don’t code outside of work': 'no_coding_after_work'
}

# Create columns boolean in nature
for status, col_name in afterwork_mapping.items():
    survey_responses[f'{col_name}'] = (survey_responses['CodingActivities'] == status)

In [125]:
survey_responses.info(120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 111 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    CodingActivities                                                                             object 
 2    EdLevel                                                                                      object 
 3    LearnCode                                                                                    object 
 4    LearnCodeOnline                                                                              object 
 5    TechDoc                                                                                      object 
 6    YearsCode                   

In [127]:
survey_responses=survey_responses.drop(columns=['CodingActivities'])

In [129]:
survey_responses.info(100)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 110 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    EdLevel                                                                                      object 
 2    LearnCode                                                                                    object 
 3    LearnCodeOnline                                                                              object 
 4    TechDoc                                                                                      object 
 5    YearsCode                                                                                    object 
 6    YearsCodePro                

Deal with EdLevel Column

In [131]:
# Define education level options
ed_levels = [
    'Primary/elementary school',
    'Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)',
    'Some college/university study without earning a degree',
    'Associate degree (A.A., A.S., etc.)',
    'Bachelor’s degree (B.A., B.S., B.Eng., etc.)',
    'Master’s degree (M.A., M.S., M.Eng., MBA, etc.)',
    'Professional degree (JD, MD, Ph.D, Ed.D, etc.)',
    'Something else'
]

# Create boolean columns
for level in ed_levels:
    # Generate clean column name
    col_name = (
        level.lower()
        .replace(' ', '_')
        .replace('/', '_')
        .replace('(', '')
        .replace(')', '')
        .replace('.', '')
        .replace(',', '')
        .replace('’', '')
        .replace('etc', '')
        .replace('e.g.', '')
        .strip('_')
    )
    
    # Create boolean column
    survey_responses[f'ed_{col_name}'] = survey_responses['EdLevel'] == level

# Deal with missing values or "NA's"
survey_responses['ed_missing'] = survey_responses['EdLevel'].isna()



In [135]:
survey_responses.info(100)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 119 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    EdLevel                                                                                      object 
 2    LearnCode                                                                                    object 
 3    LearnCodeOnline                                                                              object 
 4    TechDoc                                                                                      object 
 5    YearsCode                                                                                    object 
 6    YearsCodePro                

In [137]:
survey_responses = survey_responses.drop(columns=['EdLevel'])

In [139]:
survey_responses.info(120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 118 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    LearnCode                                                                                    object 
 2    LearnCodeOnline                                                                              object 
 3    TechDoc                                                                                      object 
 4    YearsCode                                                                                    object 
 5    YearsCodePro                                                                                 object 
 6    DevType                     

Deal with column 'LearnCode'

In [141]:


# Define learning methods (from your list)
learn_methods = [
    'Books / Physical media',
    'Coding Bootcamp',
    'Colleague',
    'Friend or family member',
    'Online Courses or Certification',
    'On the job training',
    'Other online resources (e.g., videos, blogs, forum, online community)',
    'School (i.e., University, College, etc)',
    'Other (please specify)'
]

# Clean method names for column creation
def clean_method_name(method):
    return (
        method.lower()
        .replace(' ', '_')
        .replace('/', '_')
        .replace('(', '')
        .replace(')', '')
        .replace('.', '')
        .replace(',', '')
        .replace('e.g.', '')
        .replace('i.e.', '')
        .strip('_')
    )

# Standardize delimiters and split responses
def split_learn_code(responses):
    if pd.isna(responses):
        return []
    # Replace semicolons with commas, then split
    return [x.strip() for x in responses.replace(';', ',').split(',')]

# Create boolean columns
for method in learn_methods:
    col_name = f"learn_{clean_method_name(method)}"
    survey_responses[col_name] = (
        survey_responses['LearnCode']
        .apply(lambda x: method in split_learn_code(x))
    )

# Handle NA values
survey_responses['learn_missing'] = survey_responses['LearnCode'].isna()





In [143]:
survey_responses.head(20)

ResponseId                                          LearnCode  \
0            1                             Books / Physical media   
1            2  Books / Physical media;Colleague;On the job tr...   
2            3  Books / Physical media;Colleague;On the job tr...   
3            4  Other online resources (e.g., videos, blogs, f...   
4            5  Other online resources (e.g., videos, blogs, f...   
5            6  School (i.e., University, College, etc);Online...   
6            7  Other online resources (e.g., videos, blogs, f...   
7            8  Other online resources (e.g., videos, blogs, f...   
8            9  Books / Physical media;Other online resources ...   
9           10  On the job training;Other online resources (e....   
10          11  Books / Physical media;Other online resources ...   
11          12  Books / Physical media;School (i.e., Universit...   
12          13  Books / Physical media;On the job training;Sch...   
13          14  Colleague;Other online resources (e.g., videos...   
14          15  Books / Physical media;Other online resources ...   
15          16  Books / Physical media;On the job training;Oth...   
16          17  Books / Physical media;Other online resources ...   
17          18  Other online resources (e.g., videos, blogs, f...   
18          19  Books / Physical media;On the job training;Oth...   
19          20  Books / Physical media;Other online resources ...   

                                      LearnCodeOnline  \
0                                                 NaN   
1   Technical documentation;Blogs;Books;Written Tu...   
2   Technical documentation;Blogs;Books;Written Tu...   
3   Stack Overflow;How-to videos;Interactive tutorial   
4   Technical documentation;Blogs;Written Tutorial...   
5                                                 NaN   
6   Technical documentation;Stack Overflow;Written...   
7   Technical documentation;Video-based Online Cou...   
8         Stack Overflow;Written-based Online Courses   
9   Technical documentation;Blogs;Written Tutorial...   
10  Technical documentation;Books;Written Tutorial...   
11                                                NaN   
12                                                NaN   
13  Technical documentation;Blogs;Stack Overflow;I...   
14  Stack Overflow;Coding sessions (live or record...   
15  Technical documentation;Books;Written Tutorial...   
16  Technical documentation;Written Tutorials;Stac...   
17  Blogs;Coding sessions (live or recorded);How-t...   
18  Technical documentation;Blogs;Books;Stack Over...   
19  Technical documentation;Blogs;Books;Written Tu...   

                                              TechDoc YearsCode YearsCodePro  \
0                                                 NaN       NaN          NaN   
1   API document(s) and/or SDK document(s);User gu...        20           17   
2   API document(s) and/or SDK document(s);User gu...        37           27   
3                                                 NaN         4          NaN   
4   API document(s) and/or SDK document(s);User gu...         9          NaN   
5                                                 NaN        10          NaN   
6                                                 NaN         7            7   
7                          First-party knowledge base         1          NaN   
8                                                 NaN        20          NaN   
9   Traditional public search engine;AI-powered se...        15           11   
10  API document(s) and/or SDK document(s);User gu...        20          NaN   
11                                                NaN        30           25   
12                                                NaN        20           12   
13  API document(s) and/or SDK document(s);User gu...        20          NaN   
14                                                NaN        15           10   
15             API document(s) and/or SDK document(s)        31           27

In [145]:
survey_responses = survey_responses.drop(columns=['LearnCode'])

In [147]:
survey_responses.info(120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 127 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    LearnCodeOnline                                                                              object 
 2    TechDoc                                                                                      object 
 3    YearsCode                                                                                    object 
 4    YearsCodePro                                                                                 object 
 5    DevType                                                                                      object 
 6    Country                     

Dealing with 'LearnCodeOnline' Column

In [149]:
# Define all learning resources
online_learn_methods = [
    'Stack Overflow',
    'Blogs',
    'Online community (including social media)',
    'Video-based online courses',
    'Online challenges (e.g., daily or weekly coding challenges)',
    'Coding sessions (live or recorded)',
    'Written-based online courses',
    'How-to videos',
    'Auditory material (e.g., podcasts)',
    'Technical documentation',
    'Certification videos',
    'Books',
    'Games that teach programming',
    'AI chatbot or other tool',
    'Written tutorials',
    'Interactive tutorial',
    'Other (please specify)'
]

# Clean method names for column creation
def clean_online_method(method):
    return (
        method.lower()
        .replace(' ', '_')
        .replace('/', '_')
        .replace('(', '')
        .replace(')', '')
        .replace('.', '')
        .replace(',', '')
        .replace('e.g.', '')
        .strip('_')
    )

# Create boolean columns for each method
for method in online_learn_methods:
    col_name = f"learn_online_{clean_online_method(method)}"
    survey_responses[col_name] = (
        survey_responses['LearnCodeOnline']
        .str.split(';')  # Split by semicolons
        .apply(lambda x: method in [item.strip() for item in x] 
               if isinstance(x, list) else False)
    )

# Handle NA values
survey_responses['learn_online_missing'] = survey_responses['LearnCodeOnline'].isna()




In [151]:
survey_responses.info(120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 145 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    LearnCodeOnline                                                                              object 
 2    TechDoc                                                                                      object 
 3    YearsCode                                                                                    object 
 4    YearsCodePro                                                                                 object 
 5    DevType                                                                                      object 
 6    Country                     

In [153]:
survey_responses = survey_responses.drop(columns=['LearnCodeOnline'])

In [155]:
survey_responses.info(120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 144 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    TechDoc                                                                                      object 
 2    YearsCode                                                                                    object 
 3    YearsCodePro                                                                                 object 
 4    DevType                                                                                      object 
 5    Country                                                                                      object 
 6    LanguageHaveWorkedWith      

Dealing with 'TechDoc' Column

In [157]:
# Define all technical documentation options
techdoc_options = [
    'API document(s) and/or SDK document(s)',
    'User guides or README files found in the source repository',
    'First-party knowledge base',
    'Traditional public search engine',
    'AI-powered search/dev tool (free)',
    'AI-powered search/dev tool (paid)',
    'Other (please specify)'
]

# Clean method names for column creation
def clean_techdoc_name(method):
    return (
        method.lower()
        .replace(' ', '_')
        .replace('/', '_')
        .replace('(', '')
        .replace(')', '')
        .replace('.', '')
        .replace(',', '')
        .replace('8/7/24, 9:43 AM', '')  # Remove date artifact if present
        .strip('_')
    )

# Create boolean columns for each documentation type
for option in techdoc_options:
    col_name = f"techdoc_{clean_techdoc_name(option)}"
    survey_responses[col_name] = (
        survey_responses['TechDoc']
        .str.split(';')  # Split by semicolons
        .apply(lambda x: option in [item.strip() for item in x] 
               if isinstance(x, list) else False)
    )

# Handle NA values
survey_responses['techdoc_missing'] = survey_responses['TechDoc'].isna()



In [159]:
survey_responses.info(120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 152 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    TechDoc                                                                                      object 
 2    YearsCode                                                                                    object 
 3    YearsCodePro                                                                                 object 
 4    DevType                                                                                      object 
 5    Country                                                                                      object 
 6    LanguageHaveWorkedWith      

In [161]:
survey_responses = survey_responses.drop(columns=['TechDoc'])

In [163]:
survey_responses.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 151 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    YearsCode                                                                                    object 
 2    YearsCodePro                                                                                 object 
 3    DevType                                                                                      object 
 4    Country                                                                                      object 
 5    LanguageHaveWorkedWith                                                                       object 
 6    LanguageWantToWorkWith      

Deal with 'YearsCode' and 'YearsCodPro' columns.  Converting from object to integer columns and dealing with string responses

In [165]:
import numpy as np

def convert_years_column(series):
    """
    Converts a years column with mixed responses to integers.
    Handles:
    - 'Less than 1 year' → 0
    - 'More than 50 years' → 51
    - Numeric strings → integers
    - NA values → np.nan
    """
    # Convert special cases first
    series = series.replace({
        'Less than 1 year': '0',
        'More than 50 years': '51'
    })
    
    # Convert to numeric, coercing errors to NaN
    return pd.to_numeric(series, errors='coerce')

# Apply to both columns
survey_responses['YearsCode'] = convert_years_column(survey_responses['YearsCode'])
survey_responses['YearsCodePro'] = convert_years_column(survey_responses['YearsCodePro'])

# Create flags for converted values
survey_responses['YearsCode_was_estimated'] = (
    survey_responses['YearsCode'].isna() & 
    survey_responses['YearsCode'].notna()
)
survey_responses['YearsCodePro_was_estimated'] = (
    survey_responses['YearsCodePro'].isna() & 
    survey_responses['YearsCodePro'].notna()
)

# Fill NA w median
code_median = survey_responses['YearsCode'].median()
codepro_median = survey_responses['YearsCodePro'].median()

survey_responses['YearsCode'] = survey_responses['YearsCode'].fillna(code_median)
survey_responses['YearsCodePro'] = survey_responses['YearsCodePro'].fillna(codepro_median)

# Convert to integer
survey_responses['YearsCode'] = survey_responses['YearsCode'].astype('Int64')  # Capital I allows NA
survey_responses['YearsCodePro'] = survey_responses['YearsCodePro'].astype('Int64')

# Verify results by counting responses and determining how many of each number of years there are
print("YearsCode value counts:")
print(survey_responses['YearsCode'].value_counts(dropna=False).head(51))
print("\nYearsCodePro value counts:")
print(survey_responses['YearsCodePro'].value_counts(dropna=False).head(51))

YearsCode value counts:
YearsCode
11    7204
10    4561
5     3723
6     3496
8     3449
7     3333
4     3290
15    2813
20    2636
12    2559
3     2518
9     2251
14    1822
25    1671
2     1642
13    1477
30    1461
16    1401
18    1228
17    1038
40     993
24     870
22     842
35     735
1      712
23     634
26     630
0      569
19     561
21     522
28     512
27     504
32     328
34     293
42     289
38     285
29     274
45     257
51     254
36     223
33     219
37     216
44     195
43     173
41     147
31     146
39     145
46      98
50      77
48      64
47      63
Name: count, dtype: Int64

YearsCodePro value counts:
YearsCodePro
7     16344
2      4168
3      4093
5      3526
10     3251
4      3215
0      2856
6      2843
1      2639
8      2549
12     1777
15     1635
20     1549
9      1493
11     1312
13     1127
14     1082
25      998
16      946
18      867
17      814
30      689
24      632
19      516
22      492
23      448
26      426
27      380
21

In [167]:
survey_responses.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 153 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    YearsCode                                                                                    Int64  
 2    YearsCodePro                                                                                 Int64  
 3    DevType                                                                                      object 
 4    Country                                                                                      object 
 5    LanguageHaveWorkedWith                                                                       object 
 6    LanguageWantToWorkWith      

In [172]:
dev_types = sorted(survey_responses['DevType'].dropna().unique())  ##create list of response options from rows while dropping NA's


Deal with column 'DevType'

In [174]:
### Convert responses from list into responses without commas consistent with other column names
def clean_devtype_name(role):
    return (
        role.lower()
        .replace(' ', '_')
        .replace(',', '')  # Remove commas only for column names
        .replace('/', '_')
        .replace('(', '')
        .replace(')', '')
        .strip('_')
    )



In [ ]:
### Create Boolean Columns
for role in dev_types:
    col_name = f"devtype_{clean_devtype_name(role)}"
    survey_responses[col_name] = (survey_responses['DevType'] == role)


### Create new column named 'devtype_missing' for Na Responses
survey_responses['devtype_missing'] = survey_responses['DevType'].isna()

In [178]:
survey_responses.info(123)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 188 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    YearsCode                                                                                    Int64  
 2    YearsCodePro                                                                                 Int64  
 3    DevType                                                                                      object 
 4    Country                                                                                      object 
 5    LanguageHaveWorkedWith                                                                       object 
 6    LanguageWantToWorkWith      

In [180]:
survey_responses = survey_responses.drop(columns=['DevType'])  #drop 'DevType' column

In [182]:
survey_responses.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 187 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    YearsCode                                                                                    Int64  
 2    YearsCodePro                                                                                 Int64  
 3    Country                                                                                      object 
 4    LanguageHaveWorkedWith                                                                       object 
 5    LanguageWantToWorkWith                                                                       object 
 6    DatabaseHaveWorkedWith      

In [184]:
### We are going to drop the column 'Country' as we do not want to take it into account as a determinant
survey_responses = survey_responses.drop(columns=['Country'])

We are also going to drop the column 'LanguageWantToWorkWith', 'DatabaseWantToWorkWth', and 'PlatformWantToWorkWith' as we don't want to take into account things that haven't been done


In [192]:
survey_responses = survey_responses.drop(columns=['LanguageWantToWorkWith'])


KeyError: "['LanguageWantToWorkWith'] not found in axis"

In [194]:
survey_responses.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 185 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    YearsCode                                                                                    Int64  
 2    YearsCodePro                                                                                 Int64  
 3    LanguageHaveWorkedWith                                                                       object 
 4    DatabaseHaveWorkedWith                                                                       object 
 5    DatabaseWantToWorkWith                                                                       object 
 6    PlatformHaveWorkedWith      

In [196]:
survey_responses=survey_responses.drop(columns=['DatabaseWantToWorkWith'])

In [198]:
survey_responses=survey_responses.drop(columns=['PlatformWantToWorkWith'])

In [202]:
survey_responses=survey_responses.drop(columns=['WebframeWantToWorkWith'])

In [204]:
survey_responses=survey_responses.drop(columns=['EmbeddedWantToWorkWith'])

In [206]:
survey_responses=survey_responses.drop(columns=['ToolsTechWantToWorkWith'])

In [210]:
survey_responses=survey_responses.drop(columns=['AIToolInterested in Using'])

In [214]:
survey_responses=survey_responses.drop(columns=['AIToolNot interested in Using'])

Continuing to eliminate responses that are not based in fact but future assumptions

In [216]:
survey_responses=survey_responses.drop(columns=['AINextNo change', 'AINextMore integrated', 'AINextLess integrated'])

In [220]:
survey_responses=survey_responses.drop(columns=['AINextMuch less integrated'])

In [222]:
survey_responses=survey_responses.drop(columns=['AINextMuch more integrated'])

In [224]:
survey_responses=survey_responses.drop(columns=['AIThreat'])  ##not impactful as previous information to impact decisions of existing professionals

In [226]:
survey_responses=survey_responses.drop(columns=['AIChallenges'])

In [228]:
survey_responses.info(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 171 columns):
 #    Column                                                                                       Dtype  
---   ------                                                                                       -----  
 0    ResponseId                                                                                   int64  
 1    YearsCode                                                                                    Int64  
 2    YearsCodePro                                                                                 Int64  
 3    LanguageHaveWorkedWith                                                                       object 
 4    DatabaseHaveWorkedWith                                                                       object 
 5    PlatformHaveWorkedWith                                                                       object 
 6    WebframeHaveWorkedWith      

In [230]:
survey_responses=survey_responses.drop(columns=['TBranch'])  ##desire to participate in future activity is not desired impactor